### ISSUE TITLE
Allow running ScyllaDB when /var/lib/scylla is an nfs mount

### ISSUE THREAD

**Issue by hiqsociety** [scylladb/scylladb#16069]
with reference to this issue:
https://github.com/scylladb/scylladb/issues/16024

i've tried all options for days and it's all not working.
pls check with latest docker used.

**Comment by mykaul** [comment]
NFS is not really a good idea. The latency is going to hurt your performance considerably. I'm not even sure about client and server side caching impact.

**Comment by hiqsociety** [comment]
@mykaul yes of course i understand the issues but i want to have it with nfs coz i'm using aws efs. can you help just provide the solution that will work irrespective of what the disk is based on?

i just need it to work.

**Comment by mykaul** [comment]
> @mykaul yes of course i understand the issues but i want to have it with nfs coz i'm using aws efs. can you help just provide the solution that will work irrespective of what the disk is based on?
> 
> i just need it to work.

I understand, not sure it's possible. See https://github.com/redpanda-data/redpanda/issues/8504#issuecomment-1408978031 for an identical issue.
What I'm unsure is - can you configure NFS for async IO?  Not sure it'd work, but it should get you past this specific issue, I hope.

**Comment by nyh** [comment]
> NFS is not really a good idea. The latency is going to hurt your performance considerably. I'm not even sure about client and server side caching impact.

I agree it's not a good idea for these reasons and perhaps others, but ideally one of two things should have happened:
1. Scylla could have exited with a clear error message saying that "Running on NFS is not supported because ...".
2. Scylla could continue to work with degredated performance or degraded reliability - and print a warning to that effect.

The error from Seastar and ugly-looking exception backtrace (see https://github.com/scylladb/scylladb/issues/16024) doesn't help the user understand what went wrong (that's option 1), and doesn't help the user to test the **functionality** of Scylla and later improve the setup (that's option 2).

Option 1 should be very easy to do, option 2 might be harder - someone will need to debug if it's a specific operation that isn't supported on NFS, or the whole async IO used in Seastar thing doesn't work with NFS - there might be a way around it (e.g., using a new backend that uses helper thread) with degraded performance. I'm not sure anyone should rush to do that, but we should at least know if this the case and not and not just rush to decide that people shouldn't even be able to **try** Scylla on NFS.

**Comment by mykaul** [comment]
I don't think distributed file systems can give Scylla the performance and potentially the semantics it needs to perform well. So it's not just NFS. I wouldn't recommend CephFS either.

**Comment by hiqsociety** [comment]
@mykaul @nyh

aws efs is working well with other dbs (nfs has intr option which i think can help with "retransmission issue". i'm sure it's solid enough for scylla.

Thank you all for the explanation BUT if async io is an issue, can i suggest perhaps we have a "cache disk" for nfs then? irrespective, i seriously doubt async io can be a huge issue with current state of hardware. we are not talking about 1 decade ago. nfs-wise, hardware is getting more powerful and especially with aws efs, it's solid. (i'm not talking about the one you will implement yourself, aws is really solid)

can anyone suggest how to "override" the failure and let it run in scylla? i dont even mind a self patch / mod version that will make this work. 
1. which line do i need to edit?
2. for scylla to have wider adoption, please enable this feature. it can also be used as a testing ground / experimental mode.

come on guys, have better faith in the underlying storage tech we can have in the modern world.

**Comment by mykaul** [comment]
> @mykaul @nyh
> 
> aws efs is working well with other dbs (nfs has intr option which i think can help with "retransmission issue". i'm sure it's solid enough for scylla.

I disagree. Random example from https://docs.openshift.com/container-platform/4.13/scalability_and_performance/recommended-performance-scale-practices/recommended-etcd-practices.html : "Avoid using the Network File System (NFS) protocol or other network based file systems."

> 
> Thank you all for the explanation BUT if async io is an issue, can i suggest perhaps we have a "cache disk" for nfs then? irrespective, i seriously doubt async io can be a huge issue with current state of hardware. we are not talking about 1 decade ago. nfs-wise, hardware is getting more powerful and especially with aws efs, it's solid. (i'm not talking about the one you will implement yourself, aws is really solid)
> 
> can anyone suggest how to "override" the failure and let it run in scylla? i dont even mind a self patch / mod version that will make this work.
> 
> 1. which line do i need to edit?
> 2. for scylla to have wider adoption, please enable this feature. it can also be used as a testing ground / experimental mode.
> 
> come on guys, have better faith in the underlying storage tech we can have in the modern world.
> 
> p.s. : i understand the performance degradation but given the options available. I'm ok with the performance degradation. efs-nfs can help save cost on maintenance without disk capacity planning ahead etc.
> 
> performance degradation of 20% or 50% is still better than most of what is available out there. so let it be. and i'm very sure with all you geniuses around, this can be further refined better than most other available solutions out there. in fact a lot of concept of scylla is much better than what i have come across (for the purpose it is designed for and of course you need to know what you are doing with it.)

Out of curiosity, why not look at EBS? While it won't perform good (enough?), we know it'll work.

**Comment by tchaikov** [comment]
> I don't think distributed file systems can give Scylla the performance and potentially the semantics it needs to perform well. So it's not just NFS. I wouldn't recommend CephFS either.

created https://github.com/scylladb/seastar/issues/1947 to track this error / limit on seastar.

**Comment by mykaul** [comment]
Thanks @tchaikov - closing as not planned for the time being.

**Comment by nyh** [comment]
@mykaul I think it's a mistake to close issues such as this. This is a real bug and the fact we don't want to spend our time to fix it doesn't mean it doesn't exist - or that other users won't encounter it - or maybe even an outside contributer might want to fix it one day.

Moreover, even if we decide we don't want to ever NFS work, I believe we at least need to fix the error message. It's a Seastar fix, but we need a Scylla issue to point to it.

**Comment by mykaul** [comment]
> @mykaul I think it's a mistake to close issues such as this. This is a real bug and the fact we don't want to spend our time to fix it doesn't mean it doesn't exist - or that other users won't encounter it - or maybe even an outside contributer might want to fix it one day.
> 
> Moreover, even if we decide we don't want to ever NFS work, I believe we at least need to fix the error message. It's a Seastar fix, but we need a Scylla issue to point to it.

It's not a bug that we don't support NFS. We also don't support running on floppy disks. It's not supported.
If anyone wants to implement an enhancement and improve it - sure, we can keep it open. I don't see why we need to keep it open for that reason - if anyone wants to implement it, they can. As well as 100,000 other configuration options and features and whatnot.
Seastar 'fix' is on its way - I'm not sure this item needs to be kept open to track this.

Anyway, I don't have a strong opinion here - I just don't like issues rotting in the backlog endlessly, and then someone has some kind of hope we might be looking at it - while we are not going to.
